### PR TITLE
Process messages immediately after is sent in chuncked streaming

### LINKED ISSUE #30690: Json stream item is not processed immediately after receiving in ChunkedStreamingMultiSubscriber.class
### Describe the bug

When using rest endpoint with `application/stream+json` content type, after receiving json object is not processed immediately. 
Client is expecting stream element separator to process element. Separator is sent as prefix with subsequent elements. This means that stream element is processed after receiving next one.

### Expected behavior

Stream element should be processed immediately after receiving.

Maybe ChunkedStreamingMultiSubscriber should add separator as postfix, not as prefix with next item.

### Actual behavior

Strem element is processed after receiving next one (or stream closed).

### How to Reproduce?

Reproducer: https://github.com/paloliska/json-stream

steps:
1. Run test in reproducer project.

### Output of `uname -a` or `ver`

Linux user 5.15.0-58-generic #64-Ubuntu SMP Thu Jan 5 11:43:13 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

### Output of `java -version`

openjdk version "17.0.5" 2022-10-18

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.16.0.Final, 3.0.1.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.6.3

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by Sgitario**
Fix https://github.com/quarkusio/quarkus/issues/30690

### CODE REVIEW

File: independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/impl/MultiInvoker.java [hunk 1]
```diff
@@ -205,12 +205,9 @@ public void handle(Buffer buffer) {
                         charset = charset == null ? "UTF-8" : charset;
                         byte[] separator = "\n".getBytes(charset);
                         int start = 0;
-                        if (startsWith(bytes, separator)) {
-                            start += separator.length;
-                        }
                         while (start < bytes.length) {
                             int end = bytes.length;
-                            for (int i = start; i < bytes.length - separator.length; i++) {
+                            for (int i = start; i < end; i++) {
```
Comment (Sgitario) [inline independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/impl/MultiInvoker.java:210]: I modified this logic to work regardless the position of the delimiter is.

File: extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/streams/StreamTestCase.java [hunk 2]
```diff
@@ -141,7 +141,6 @@ public void testStreamJsonMultiFromMulti() {
 
     private void testJsonMulti(String path) {
         Client client = ClientBuilder.newBuilder().register(new JacksonBasicMessageBodyReader(new ObjectMapper())).build();
-        ;
```
Comment (Sgitario) [inline extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/streams/StreamTestCase.java:144]: This is unrelated to the changes, but it was a leftover

**Comment by geoand** [comment]
@FroMage mind taking a look as well?

File: extensions/resteasy-reactive/rest-client-reactive/deployment/src/test/java/io/quarkus/rest/client/reactive/StreamJsonTest.java [hunk 3]
```diff
@@ -109,6 +114,21 @@ void shouldReadNdjsonFromSingleMessage() throws InterruptedException {
...
+    @Test
+    public void shouldReadUpToThreeTicks() {
+        createClient(uri)
+                .ticks()
+                .onItem()
+                .invoke(Objects::nonNull)
+                .subscribe()
+                .withSubscriber(AssertSubscriber.create(3))
+                // wait for 3 ticks plus some half tick ms of extra time (this should not be necessary, but CI is slow)
+                .awaitItems(3, Duration.ofMillis((TICK_EVERY_MS * 3) + (TICK_EVERY_MS / 2)));
```
Comment (paloliska) [inline extensions/resteasy-reactive/rest-client-reactive/deployment/src/test/java/io/quarkus/rest/client/reactive/StreamJsonTest.java:129]: Should be duration of `(TICK_EVERY_MS * 2) ...` as 1st item is issued at time 0, 2nd at time 200ms, 3rd at time 400ms. So wait time 200ms * 2,5 should be correct

Comment (Sgitario) [reply extensions/resteasy-reactive/rest-client-reactive/deployment/src/test/java/io/quarkus/rest/client/reactive/StreamJsonTest.java:129]: This does not have to be precise taking into account that there are more things involved (REST call, subscriptions, etc). 
According to my experience, 200 ms of back time is enough to reproduce the issue and verify this is indeed fixed.

**Review by geoand — APPROVED** [review]

**Comment by paloliska** [comment]
Will this be also backported to 2.16.x branch?

**Comment by geoand** [comment]
@FroMage please review this

**Comment by paloliska** [comment]
Is it possible to merge this, please?

**Comment by paloliska** [comment]
@FroMage could you please check this?
or @geoand  is it possible to merge this with just 1 approval?

**Review by FroMage — COMMENTED** [review]
It sounds fair to send a newline after each element… except that we don't know if there will be a next one, at this stage.
So won't a stream of one string be indistinguishable from a stream of one string plus one empty string, from the client's perspective?

**Comment by Sgitario** [comment]
> It sounds fair to send a newline after each element… except that we don't know if there will be a next one, at this stage. So won't a stream of one string be indistinguishable from a stream of one string plus one empty string, from the client's perspective?

Tbh, I don't follow what you're asking or requesting as per your comment. Is there a use case you foresee that these changes won't work? In terms of functionality, this will work the same as before, the only thing that changes is when to process the message.

**Comment by paloliska** [comment]
@FroMage do you need some input or help to test this PR? I could make some additional tests if you have any scenarios in mind.

**Comment by paloliska** [comment]
issue https://github.com/quarkusio/quarkus/issues/30690 updated with reproducer for quarkus 3.0.1.Final

**Comment by Sgitario** [comment]
> issue #30690 updated with reproducer for quarkus 3.0.1.Final

Thanks! @FroMage any updates on this?

**Comment by geoand** [comment]
@FroMage can you have a look at this one please?

**Review by FroMage — APPROVED** [review]
Sorry for the delay.

It's the names that are super confusing.

My concern was whether we would be able from the client's perspective to deal with adding `\n` at the end of each element, even when we don't know there's a next one.

Consider the following data sent and how they're received:

- Sending nothing → client receives nothing, that's a 0-element reception of no chunks
- Sending "foo" → client receives one element with data `foo`
- Sending "foo\nbar" → client receives two elements with data `foo` and `bar`
- Sending "foo\n" → client receives two elements with data `foo` and `` (the empty string)

So, previously, `\n` was a separator between elements, and now it's a line terminator. This could lead to bad issues.

Now, it happens that we're streaming differently in different cases:

- `APPLICATION_NDJSON_TYPE` and `APPLICATION_STREAM_JSON_TYPE` will stream using `ChunkedStreamingMultiSubscriber`, and since both types are JSON types, we can only stream JSON values, where Strings are delimited by `""` and so a trailing empty line will never be considered as a valid value, and hopefully will be discarded by the client. I haven't checked the specs, though, so I'm not sure it's even valid.
- Any other streaming case that isn't SSE will use `StreamingMultiSubscriber` (which is a supertype of `ChunkedStreamingMultiSubscriber` and frankly that is confusing, because the names are just confusing), which allows JSON or regular HTTP streaming. Now, in this case, adding `\n` as a line terminator would be very problematic, especially for regular HTTP streaming. But this class doesn't add line separators in this PR, so it's fine.

Frankly, this is much more complex and delicate than it appears at first by just glancing at this PR. I haven't checked the two specs whose behaviour this modifies and I hope it won't affect clients with a trailing empty chunk. I am reasonably confident it should not, given that JSON doesn't allow empty values and would consider it as whitespace, but I hope a JS client would not produce an error. I hope the PR authors checked the specs :)

**Comment by Sgitario** [comment]
> 
> Consider the following data sent and how they're received:
> 
> * Sending nothing → client receives nothing, that's a 0-element reception of no chunks
> * Sending "foo" → client receives one element with data `foo`
> * Sending "foo\nbar" → client receives two elements with data `foo` and `bar`
> * Sending "foo\n" → client receives two elements with data `foo` and `` (the empty string)
> 

I've just checked that when sending `foo\n`, the client will only receive one element, so I think we're fine here.

**Comment by FroMage** [comment]
> I've just checked that when sending `foo\n`, the client will only receive one element, so I think we're fine here.

Which client? With what mime type? Using chunked encoding?

**Comment by Sgitario** [comment]
> > I've just checked that when sending `foo\n`, the client will only receive one element, so I think we're fine here.
> 
> Which client? With what mime type? Using chunked encoding?

REST Client Reactive:
```java
@GET
@Path("/stream/string")
@Produces(RestMediaType.APPLICATION_STREAM_JSON)
@RestStreamElementType(MediaType.APPLICATION_JSON)
Multi<String> readString();
```

Server side:
```java
@Path("/stream")
public static class StreamingResource {
    @GET
    @Path("/string")
    @Produces(RestMediaType.APPLICATION_STREAM_JSON)
    @RestStreamElementType(MediaType.APPLICATION_JSON)
    public String readString() {
        return "\"one\"\n\"two\"\n\"3\"\n\"four\"\n";
    }
```

**Comment by FroMage** [comment]
Yeah, those are not the clients I was most worried about. But I guess we'll hear about JS clients if it breaks :)

**Comment by FroMage** [comment]
> Oh, I see now, sorry for my misunderstanding!

My fault, I often make answers too short to properly understand, sorry :)

**Comment by Sgitario** [comment]
The test failures are unrelated. Merging.

**Pull request merged**